components/datetime/src/skeleton.rs

```diff
@@ -263,10 +291,339 @@ impl From<fields::SymbolError> for SkeletonError {
+// - MMM ≈ MM    (Sep ≈ 09)
+//   MMM
+const TEXT_VS_NUMERIC_DISTANCE: u16 = 10;
+
+// D. Symbols representing substantial differences (week of year vs week of month) are given much
+// larger a distances from each other.
+// - d ≋ D;     (12 ≋ 345) Day of month vs Day of year
+const SUBSTANTIAL_DIFFERENCES_DISTANCE: u16 = 100;
+
+// Finally, missing or extra symbols are the biggest distance.
+const MISSING_OR_EXTRA_SYMBOL_DISTANCE: u16 = 1000;
+
+/// According to the [UTS 35 skeleton matching algorithm](https://unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons)
+/// there will be a guaranteed match for a skeleton. However, with this initial implementation,
+/// there is no attempt to add on missing fields. This enum encodes the variants for the current
+/// search for a best skeleton.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum BestSkeleton<'a> {
+    AllFieldsMatch(AvailableFormatPattern<'a>),
+    MissingOrExtraFields(AvailableFormatPattern<'a>),
+    NoMatch,
+}
+
+/// A partial implementation of the [UTS 35 skeleton matching algorithm](https://unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons).
+///
+/// The following is implemented:
+///
+///  * Compute a score based on the best possible match for the given fields.
+///  * Select the skeleton with highest score.
+///
+/// The following is not implemented:
+///
+///  * 2.6.2.1 Matching Skeletons
+///    - TODO(#584) - Modify the resulting pattern to have fields of the same length. For example requesting
+///      a skeleton "yMMMMd" can have a best match of ["yMMMd", "d MMM y"]. This pattern should
+///      then be modified to use the requested length to produce a pattern "d MMMM y".
+///      However, fields should not be changed from numeric to text.
+///  * 2.6.2.2 Missing Skeleton Fields
+///    - TODO(#585) - The mechanism to combine a date pattern and a time pattern.
+///    - TODO(#586) - Using the CLDR appendItems field. Note: There is not agreement yet on how
+///      much of this step to implement. See the issue for more information.
+pub fn get_best_available_format_pattern<'a>(
+    available_format_patterns: impl Iterator<Item = AvailableFormatPattern<'a>> + 'a,
+    fields: &[Field],
+) -> BestSkeleton<'a> {
+    let mut closest_format_pattern = None;
+    let mut closest_distance: u16 = u16::MAX;
+    let mut closest_missing_fields = 0;
+
+    for available_format_pattern in available_format_patterns {
+        let skeleton = &available_format_pattern.skeleton;
+        debug_assert!(
+            skeleton.fields_len() <= MAX_FIELDS as usize,
+            "The distance mechanism assumes skeletons are less than MAX_FIELDS in length."
+        );
+        let mut missing_fields = 0;
+        let mut distance: u16 = 0;
+        // The distance should fit into a u16.
+        #[cfg(test)]
+        static_assertions::const_assert!(MAX_FIELDS * MISSING_OR_EXTRA_SYMBOL_DISTANCE < u16::MAX);
```

Comment:
Issue: In the algorithm below, you have `distance += MISSING_OR_EXTRA_SYMBOL_DISTANCE` for mismatched requested *or* mismatched skeleton fields.  So, you could have a request like "GyQMdEz" and a skeleton like "hmsSa" and you exceed your max fields limit of 10.